src/core/server/plugins/plugin.test.ts

```diff
@@ -0,0 +1,213 @@
+  ).rejects.toMatchInlineSnapshot(
+    `[Error: Plugin "some-plugin-id" does not export "plugin" definition (plugin-without-initializer-path).]`
+  );
+});
+
+test('`start` fails if plugin initializer is not a function', async () => {
+  const manifest = createPluginManifest();
+  const plugin = new Plugin(
+    'plugin-with-wrong-initializer-path',
+    manifest,
+    createPluginInitializerContext(coreContext, manifest)
+  );
+
+  await expect(
+    plugin.start(createPluginStartContext(coreContext, plugin), {})
+  ).rejects.toMatchInlineSnapshot(
+    `[Error: Definition of plugin "some-plugin-id" should be a function (plugin-with-wrong-initializer-path).]`
+  );
+});
+
+test('`start` fails if initializer does not return object', async () => {
+  const manifest = createPluginManifest();
+  const plugin = new Plugin(
+    'plugin-with-initializer-path',
+    manifest,
+    createPluginInitializerContext(coreContext, manifest)
+  );
+
+  mockPluginInitializer.mockReturnValue(null);
+
+  await expect(
+    plugin.start(createPluginStartContext(coreContext, plugin), {})
+  ).rejects.toMatchInlineSnapshot(
+    `[Error: Initializer for plugin "some-plugin-id" is expected to return plugin instance, but returned "null".]`
+  );
+});
+
+test('`start` fails if object returned from initializer does not define `start` function', async () => {
+  const manifest = createPluginManifest();
+  const plugin = new Plugin(
+    'plugin-with-initializer-path',
+    manifest,
+    createPluginInitializerContext(coreContext, manifest)
+  );
+
+  const mockPluginInstance = { run: jest.fn() };
+  mockPluginInitializer.mockReturnValue(mockPluginInstance);
+
+  await expect(
+    plugin.start(createPluginStartContext(coreContext, plugin), {})
+  ).rejects.toMatchInlineSnapshot(
+    `[Error: Instance of plugin "some-plugin-id" does not define "start" function.]`
+  );
+});
+
+test('`start` initializes plugin and calls appropriate lifecycle hook', async () => {
+  const manifest = createPluginManifest();
+  const initializerContext = createPluginInitializerContext(coreContext, manifest);
+  const plugin = new Plugin('plugin-with-initializer-path', manifest, initializerContext);
+
+  const mockPluginInstance = { start: jest.fn().mockResolvedValue({ contract: 'yes' }) };
+  mockPluginInitializer.mockReturnValue(mockPluginInstance);
+
+  const startContext = createPluginStartContext(coreContext, plugin);
+  const startDependencies = { 'some-required-dep': { contract: 'no' } };
+  await expect(plugin.start(startContext, startDependencies)).resolves.toEqual({ contract: 'yes' });
+
+  expect(mockPluginInitializer).toHaveBeenCalledTimes(1);
+  expect(mockPluginInitializer).toHaveBeenCalledWith(initializerContext);
+
+  expect(mockPluginInstance.start).toHaveBeenCalledTimes(1);
+  expect(mockPluginInstance.start).toHaveBeenCalledWith(startContext, startDependencies);
+});
+
+test('`stop` does nothing if plugin is not started', async () => {
```

Comment:
Perhaps `stop` should throw an error if it's invoked when the plugin hasn't been started. Thoughts?

Reply:
Yeah, thanks, I'll change it. I think it makes sense since we can't accidentally call `stop` on not started plugin, but if we do it's a bug and we'd better know about that early.